model/vm_pool.rb

```diff
@@ -5,21 +5,22 @@
 class VmPool < Sequel::Model
   one_to_one :strand, key: :id
   one_to_many :vms, key: :pool_id
+  one_to_many :idle_vms, key: :pool_id, class: Vm, conditions: {tainted_at: nil, display_state: "running"}
 
   include ResourceMethods
 
   include SemaphoreMethods
   semaphore :destroy
 
-  def pick_vm
-    DB.transaction do
-      # first lock the whole pool in the join table so that no other thread can
-      # pick a vm from this pool
-      vms_dataset.for_update.all
-      pick_vm_id_q = vms_dataset.left_join(:github_runner, vm_id: :id)
-        .where(Sequel[:github_runner][:vm_id] => nil, Sequel[:vm][:display_state] => "running")
-        .select(Sequel[:vm][:id])
-      Vm.where(id: pick_vm_id_q).first
-    end
+  def pick_vm(new_name)
+    vm = idle_vms_dataset.order(Sequel.lit("RANDOM()")).first
+    return if vm.nil?
+
+    # If we can successfully update the VM, we can be certain that no other
+    # thread has selected it. If the update fails, we continue to pick different
+    # VMs until we find an idle one or exhaust the pool of runners.
+    return pick_vm(new_name) if idle_vms_dataset.where(id: vm.id).update(name: new_name, tainted_at: Time.now).zero?
+
+    vm
```

Comment:
Something about this code seems a little weird, it's odd to:

1. Run a query that qualifies a tuple
2. Run it again, but this time with update

Particularly when there's not much going on between steps (1) and (2).

How about `update` with `returning`? You can use `Model#call` to make model objects with any hash.

Reply:
The challenge is updating a single idle vm without an ID for filtering in the update query. Using `limit` with `update` results in `Sequel::InvalidOperation: Dataset#update not supported on datasets with limits or offsets`.
A potential solution is to incorporate the first query as a subquery in the second one.

An example:
```
p.idle_vms_dataset
    .where(id: p.idle_vms_dataset.order(Sequel.lit("RANDOM()")).limit(1).select(:id))
    .update(name: "new-name", tainted_at: Time.now)
```

Reply:
you could also use a query with a CTE, in SQL it'd look like:

```sql
with candidate as (
  select id from vm where ... limit 1
)
update vm set ... from candidate where vm.id = candidate.id
returning ...
```